library/core/src/ptr/non_null.rs

```diff
@@ -372,13 +372,13 @@ impl<T: ?Sized> NonNull<T> {
     ///
     /// [the module documentation]: crate::ptr#safety
     #[stable(feature = "nonnull", since = "1.25.0")]
-    #[rustc_const_unstable(feature = "const_ptr_as_ref", issue = "91822")]
+    #[rustc_const_stable(feature = "const_nonnull_as_ref", since = "CURRENT_RUSTC_VERSION")]
     #[must_use]
     #[inline]
     pub const unsafe fn as_ref<'a>(&self) -> &'a T {
         // SAFETY: the caller must guarantee that `self` meets all the
         // requirements for a reference.
-        unsafe { &*self.as_ptr() }
+        unsafe { &*self.as_ptr().cast_const() }
```

Comment:
Why is the `cast_const` needed now?

Reply:
Without the cast it will fail to compile with `error[E0658]: dereferencing raw mutable pointers in constant functions is unstable` [(playground)](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=6029f1cf0b42fe2f524032eab93908cc)